### ISSUE TITLE
Inefficient code generated for `__builtin_shufflevector` with runtime mask on ARM

### ISSUE THREAD

**Issue by lawben** [llvm/llvm-project#60515]
I'm currently looking at using `__builtin_shufflevector(vec, mask)` in a project on an M1 MacBook and I found that the code generated is worse compared to a) the x86 version and b) what it could be. Note: This is not the version with a constant mask but with a runtime mask. This is technically not documented but it is implemented. There is an [open issue](https://github.com/llvm/llvm-project/issues/59678) for the documentation side.

Here is a [godbolt link](https://godbolt.org/z/4fPKaKz8M) for the code example.

```cpp
using VecT __attribute__((vector_size(16))) = uint8_t;

VecT shuffle_builtin(VecT vec, VecT mask) {
    return __builtin_shufflevector(vec, mask);
}
```

This code is compiled to two instructions on x86 (`vpand` + `vpshufb`). The `vpand` is not strictly necessary, but it get's created in the frontend and is not removed during instruction selection. However, on ARM with NEON, this gets compiled to 66 instructions, logically consisting of 16x "extract element at i and insert it". The same applies to vectors with 2/4/8 elements [(godbolt)](https://godbolt.org/z/3aeqeqPv6). 

I'm currently looking into creating a patch for this for 16 elements, as this translates directly to a NEON `TBL` instruction. As this pattern is detected during instruction selection on x86, I guess it can also be detected for aarch64. Handling 2/4/8/ elements requires some additional thought, but I assume this can also be modeled in a similar way.

**Comment by RKSimon** [comment]
@davemgreen @lawben Resolve this after 267d6d665cf2379ebfcc65fa385a35529c83a7d0?

**Comment by davemgreen** [comment]
It looks like it's working well for v16i8, but not for other sizes yet. (From https://godbolt.org/z/3aeqeqPv6).

They should hopefully be able to use a generalization of the same code.

**Comment by lawben** [comment]
There is some logic for this in [`AArch64TargetLowering::ReconstructShuffle`](https://github.com/llvm/llvm-project/blob/179d2efce6dea2caddb56f8bcba20b62dc5b91cb/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L10468C9-L10468C50) for compile-time immediates in the shuffle. This would need to be copied and adapted to work with runtime values, i.e., there need to be some multiplications and additions to get the correct byte offsets.

For a project I worked on, I played around with this a bit already. You can see some of the logic [here (godbolt)](https://godbolt.org/z/oT8r3n6f6). I think it makes sense to extend this to vectors with 16-bit elements. For 32- and 64-bit entries, I'm not sure if the overhead is worth it (at least in my current approach. There may be a better way to do this).
